File: backend/models/user_context.py
"""
User Context Model
Contains user preferences and constraints for recipe recommendations
"""
from pydantic import BaseModel, Field
from typing import Optional, List


class UserContext(BaseModel):
    """
    User context for personalized recipe recommendations

    This model combines:
    - Diet preferences (vegetarian, vegan, gluten-free, keto)
    - Allergen restrictions (hard filter - recipes with allergens are excluded)
    - Cuisine preferences (Turkish, Italian, Asian, Mediterranean)
    - Available ingredients from user's inventory
    - Cooking time and calorie constraints
    """

    diet_types: List[str] = Field(
        default=[],
        description="Diet types: Vejetaryen, Vegan, Glutensiz, Keto"
    )

    allergens: List[str] = Field(
        default=[],
        description="Allergen categories: Fındık/Fıstık, Süt, Yumurta, Deniz Ürünleri, Gluten"
    )

    cuisines: List[str] = Field(
        default=[],
        description="Preferred cuisines: Türk, İtalyan, Asya, Akdeniz"
    )

    available_ingredients: List[str] = Field(
        default=[],
        description="Ingredients available in user's inventory"
    )

    max_cooking_time: Optional[int] = Field(
        default=None,
        description="Maximum cooking time in minutes (null = no limit)"
    )

    max_calories: Optional[int] = Field(
        default=None,
        description="Maximum calories per serving (null = no limit)"
    )

    class Config:
        json_schema_extra = {
            "example": {
                "diet_types": ["Vejetaryen"],
                "allergens": ["Fındık/Fıstık", "Süt"],
                "cuisines": ["Türk", "Akdeniz"],
                "available_ingredients": ["domates", "soğan", "zeytinyağı", "pirinç"],
                "max_cooking_time": 30,
                "max_calories": 500
            }
        }

    def has_restrictions(self) -> bool:
        """Check if user has any dietary restrictions"""
        return bool(self.diet_types or self.allergens)

    def has_preferences(self) -> bool:
        """Check if user has any preferences set"""
        return bool(
            self.diet_types or
            self.allergens or
            self.cuisines or
            self.max_cooking_time or
            self.max_calories
        )

    def to_system_prompt(self) -> str:
        """
        Generate system prompt section for LLM/RAG queries

        Returns:
            Formatted string describing user's dietary context
        """
        parts = []

        if self.diet_types:
            diets = ", ".join(self.diet_types)
            parts.append(f"Kullanıcının diyet tercihleri: {diets}")

        if self.allergens:
            allergens = ", ".join(self.allergens)
            parts.append(
                f"⚠️ UYARI: Kullanıcının şu alerjileri var: {allergens}. "
                f"Bu malzemeleri ve türevlerini KESİNLİKLE içeren tarifleri önerme!"
            )

        if self.cuisines:
            cuisines = ", ".join(self.cuisines)
            parts.append(f"Tercih edilen mutfaklar: {cuisines}")

        if self.available_ingredients:
            ingredients = ", ".join(self.available_ingredients[:10])
            parts.append(f"Kullanıcının mevcut malzemeleri: {ingredients}")

        if self.max_cooking_time:
            parts.append(f"Maksimum pişirme süresi: {self.max_cooking_time} dakika")

        if self.max_calories:
            parts.append(f"Maksimum kalori limiti: {self.max_calories} kcal")

        if not parts:
            return ""

        return "\n".join(parts)

File: android-app/app/src/main/java/com/yemekonerisistemi/app/data/model/UserContext.kt
package com.yemekonerisistemi.app.data.model

import com.google.gson.annotations.SerializedName

/**
 * User context model for personalized recipe recommendations
 * Sent to backend API to apply filtering and customize LLM responses
 */
data class UserContext(
    @SerializedName("diet_types")
    val dietTypes: List<String> = emptyList(),

    @SerializedName("allergens")
    val allergens: List<String> = emptyList(),

    @SerializedName("cuisines")
    val cuisines: List<String> = emptyList(),

    @SerializedName("available_ingredients")
    val availableIngredients: List<String> = emptyList(),

    @SerializedName("max_cooking_time")
    val maxCookingTime: Int? = null,

    @SerializedName("max_calories")
    val maxCalories: Int? = null,

    @SerializedName("cooking_time_preferences")
    val cookingTimePreferences: List<String> = emptyList(),

    @SerializedName("calorie_preferences")
    val caloriePreferences: List<String> = emptyList(),

    @SerializedName("skill_levels")
    val skillLevels: List<String> = emptyList(),

    @SerializedName("spice_preferences")
    val spicePreferences: List<String> = emptyList(),

    @SerializedName("serving_sizes")
    val servingSizes: List<String> = emptyList(),

    @SerializedName("meal_types")
    val mealTypes: List<String> = emptyList()
) {
    /**
     * Check if user has any dietary restrictions (allergens or diet types)
     */
    fun hasRestrictions(): Boolean = allergens.isNotEmpty() || dietTypes.isNotEmpty()

    /**
     * Check if user has any preferences set
     */
    fun hasPreferences(): Boolean = dietTypes.isNotEmpty() ||
            allergens.isNotEmpty() ||
            cuisines.isNotEmpty() ||
            maxCookingTime != null ||
            maxCalories != null ||
            cookingTimePreferences.isNotEmpty() ||
            caloriePreferences.isNotEmpty() ||
            skillLevels.isNotEmpty() ||
            spicePreferences.isNotEmpty() ||
            servingSizes.isNotEmpty() ||
            mealTypes.isNotEmpty()

    companion object {
        /**
         * Create an empty UserContext with no preferences
         */
        fun empty(): UserContext = UserContext()
    }
}
